incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -91,9 +104,58 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 	// TODO report the "SubnamespaceConflict" in the HNS reconciliation. See issue:
 	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/490
 
+	// Make sure finalizers are set before writing the instance.
+	inst.ObjectMeta.Finalizers = []string{api.MetaGroup}
 	return ctrl.Result{}, r.writeInstance(ctx, log, inst)
 }
 
+func (r *HierarchicalNamespaceReconciler) syncDeleting(log logr.Logger, inst *api.HierarchicalNamespace) (updateFinalizers bool) {
+	r.forest.Lock()
+	defer r.forest.Unlock()
+
+	state := inst.Status.State
+	cnm := inst.Name
+	pnm := inst.Namespace
+	cns := r.forest.Get(cnm)
+	updateFinalizers = false
+
+	switch {
+	// If the owned namespace doesn't exist, remove the finalizer. The missing state
+	// is excluded here in case it's being created by the HC reconciler concurrently.
```

Comment:
This will get much simpler once HNSR creates/deletes the namespace :) Feel free to take some shortcuts for now and comment that fact.